### ISSUE TITLE
AMQP binding does not retry appropriately.

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#288]
UPDATE: this description is wrong. older versions didn't work, this is not a regression.
see later entries.

in prior versions of Sarracenia (v2.19.01b1) when attempting to bind, if there was a failure to bind, the application would correctly loop attempting to get good bindings:

```

2019-12-14 08:50:31,012 [INFO] sr_report run
2019-12-14 08:50:31,013 [INFO] AMQP  broker(localhost) user(tfeed) vhost(/)
2019-12-14 08:50:31,013 [ERROR] AMQP Sender cannot connect to: localhost
2019-12-14 08:50:31,013 [ERROR] Type=<class 'TypeError'>, Value=port must be an int, but got None
2019-12-14 08:50:31,013 [ERROR] Sleeping 5 seconds ...
^C2019-12-14 08:50:35,793 [INFO] signal stop (SIGINT)
2019-12-14 08:50:35,793 [INFO] sr_report stop

```

The message isn't terribly helpful, but the behaviour is.  it loops until the binding succeeds.
in later versions, this happens:
```

019-12-14 08:52:28,248 [INFO] sr_report run
2019-12-14 08:52:28,248 [INFO] AMQP  broker(localhost) user(tfeed) vhost(/)
2019-12-14 08:52:28,248 [INFO] Using amqp module (AMQP 0-9-1)
2019-12-14 08:52:28,255 [INFO] Binding queue q_tfeed.sr_report.tsarra_f20.16806016.42268017 with key v02.post.# from exchange xsarra on broker amqp://tfeed@localhost/
2019-12-14 08:52:28,257 [ERROR] bind queue: q_tfeed.sr_report.tsarra_f20.16806016.42268017 to exchange: xsarra with key: v02.post.# failed with Queue.bind: (404) NOT_FOUND - no exchange 'xsarra' in vhost '/'
2019-12-14 08:52:28,257 [ERROR] Permission issue with tfeed@localhost or exchange xsarra not found.
2019-12-14 08:52:28,257 [WARNING] this process will not receive pulse message
2019-12-14 08:52:28,257 [INFO] reading from to tfeed@localhost, exchange: xsarra
2019-12-14 08:52:28,258 [INFO] report_back suppressed
2019-12-14 08:52:28,258 [INFO] sr_retry on_heartbeat
201

```

The bind fails, but it continues on, and does not try to bind. Result:  the application never tries to bind again, and therefore never receives any messages.   This results in many failures related to timing of component startup.  if the components are started in the wrong order, then the exchange isn't there art the right time for the subscribers to use them, and they don't connect.  This requires restarting components later, or being very carefuly about startup order.

The correct thing for the code to do is not return until a successful bind has occurred, which is what it used to do.

**Comment by petersilva** [comment]
it is broken as of v2.19.02b2

**Comment by petersilva** [comment]
after more testing... not sure if this is a regression or not... I'm testing older versions, and seeing the same behaviour.  which is odd.  I got confused by the amqplib/pika/amqp library transition work. it looks like it was only looping with pika, and amqplib has apparently never worked... hmm... this isn't desirable behaviour.

**Comment by petersilva** [comment]
binding looping with exponential backoff implemented here: e11aa6f360dcc1f4be34b2f234030502c9623cd9

**Comment by petersilva** [comment]
Just in case someone is relying on bind failing and continuing... added *subtopic None*
to suppress setting bindings in consumers.  This means there is no risk to the loop behaviour being added. ( 8d4728367b1e341ab6cd8499380943dd316319f1 )

**Comment by petersilva** [comment]
released in v2.19.12bsomething...
